Task 3/Task 3.2/Task3.2.1.Library/DynamicArray.cs

```diff
@@ -0,0 +1,281 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Task3._2._1.Library.Enumerators;
+
+namespace Task3._2._1.Library
+{
+    /*
+     * На базе массива (именно массива, специфичные коллекции .NET не использовать) реализовать
+     * свой собственный класс DynamicArray<T>, представляющий собой массив с запасом, хранящий
+     * объекты произвольных типов 
+     */
+    public class DynamicArray<T> : IEnumerable, IEnumerable<T>, ICloneable
+    {
+        private T[] _array;
+
+        /// <summary>
+        /// Свойство Length — получение количества элементов.
+        /// </summary>
+        public int Length { get; private set; } // 8
+
+        /// <summary>
+        /// Свойство Capacity — получение ёмкости: длины внутреннего массива.
+        /// </summary>
+        public int Capacity { get => _array.Length; } // 9
+
+        /// <summary>
+        /// Конструктор без параметров (создаётся массив ёмкостью 8 элементов).
+        /// </summary>
+        public DynamicArray() // 1
+        {
+            _array = new T[8];
+            Length = 0;
+        }
+
+        /// <summary>
+        /// Конструктор с одним целочисленным параметром (создаётся массив указанной ёмкости).
+        /// </summary>
+        /// <param name="capacity"></param>
+        public DynamicArray(int capacity) // 2
+        {
+            _array = new T[capacity];
+            Length = 0;
+        }
+
+        /// <summary>
+        /// Конструктор, который в качестве параметра принимает коллекцию, реализующую 
+        /// интерфейс IEnumerable<T>, создаёт массив нужного размера и копирует в него все
+        /// элементы из коллекции.
+        /// </summary>
+        public DynamicArray(IEnumerable<T> source)  // 3
+        {
+            _array = source.ToArray();
+            Length = source.Count();
+        }
+
+
+        /// <summary>
+        /// The method must be called before each addition to the array.
+        /// Increases the capacity to the required value, if needed.
+        /// (The array will grow as well)
+        /// </summary>
+        private void IncreaseCapacity(int requiredValue) // 2*
+        {
+            if (requiredValue <= Capacity)
+                return;
+
+            int newCapacity = Capacity == 0 ? 8 : Capacity;
+
+            while (newCapacity < requiredValue) { newCapacity *= 2; }
```

Comment:
Why not calculate multiplier instead of using `while`? Or just use proposed capacity with some small additional buffer?